### PR TITLE
chore(backend): clean up incoming payment JSON conversion

### PR DESCRIPTION
**Pull request by wilsonianb**
## Changes proposed in this pull request

- add incoming payment `url` and `completed` getters
- use undefined in incoming payment toJSON

## Context

- setup for #581 

## Checklist

- [ ] Related issues linked using `fixes #number`
- [x] Tests added/updated
- [ ] Documentation added
- [x] Make sure that all checks pass

### CODE REVIEW

File: packages/backend/migrations/20220228162503_create_incoming_payments_table.js [hunk 1]
```diff
@@ -12,8 +12,6 @@ exports.up = function (knex) {
     table.string('externalRef').nullable()
     table.uuid('connectionId').notNullable()
 
-    table.uuid('assetId').notNullable()
-    table.foreign('assetId').references('assets.id')
```
Comment (wilsonianb) [inline packages/backend/migrations/20220228162503_create_incoming_payments_table.js:16]: :confounded: I forgot about the _possibility_ of a payment pointer's asset changing, and thus the need to store the asset on the payment...

File: packages/backend/src/connector/core/lib/crypto.ts [hunk 2]
```diff
@@ -1,16 +1,8 @@
...
   return createHash('sha256').update(preimage).digest()
 }
-export function uuid(): string {
-  const random = randomBytes(16)
-  random[6] = (random[6] & 0x0f) | 0x40
-  random[8] = (random[8] & 0x3f) | 0x80
-  return random
-    .toString('hex')
-    .replace(/(.{8})(.{4})(.{4})(.{4})(.{12})/, '$1-$2-$3-$4-$5')
-}
```
Comment (sabineschaller) [inline packages/backend/src/connector/core/lib/crypto.ts:13]: We used to have our own uuid function? 🤯

File: packages/backend/src/open_payments/payment/incoming/model.ts [hunk 3]
```diff
@@ -197,14 +212,14 @@ export class IncomingPayment
             ...json.incomingAmount,
             value: json.incomingAmount.value.toString()
           }
-        : null,
+        : undefined,
```
Comment (sabineschaller) [inline packages/backend/src/open_payments/payment/incoming/model.ts:215]: `undefined` is not proper JSON, but `null` is. I'd prefer keeping it that way. Same goes for lines 221 and 222.

File: packages/backend/src/open_payments/payment/incoming/model.ts [hunk 4]
```diff
@@ -216,11 +231,11 @@ export class IncomingPayment
...
   id: string
   paymentPointer: string
-  incomingAmount: AmountJSON | null
+  incomingAmount?: AmountJSON
   receivedAmount: AmountJSON
   completed: boolean
-  description: string | null
-  externalRef: string | null
+  description?: string
+  externalRef?: string
```
Comment (sabineschaller) [inline packages/backend/src/open_payments/payment/incoming/model.ts:238]: Same as above.

**Review by sabineschaller — COMMENTED** [review]
I can already see all the merge conflicts that will give me ... But great improvements 😉

File: packages/backend/src/open_payments/payment/incoming/model.ts (hunk 3 above)
Comment (wilsonianb) [reply packages/backend/src/open_payments/payment/incoming/model.ts]: You've brought this up before, and I forgot.
I dropped the `null` to `undefined` commit.

I might keep exploring a comparable change though. My intent is to have representations of the schemas in our OpenAPI spec, and `null` is distinct from a property not being required.
https://json-schema.org/understanding-json-schema/reference/null.html
https://json-schema.org/understanding-json-schema/reference/object.html#required

Comment (wilsonianb) [reply packages/backend/src/open_payments/payment/incoming/model.ts]: https://www.typescriptlang.org/tsconfig#exactOptionalPropertyTypes :eyes:

Comment (wilsonianb) [reply packages/backend/src/open_payments/payment/incoming/model.ts]: - https://github.com/interledger/rafiki/issues/594

**Review by dclipp — APPROVED** [review]

**Review by sabineschaller — APPROVED** [review]

**Pull request merged**